src/gluonts/dataset/arrow/dec.py

```diff
@@ -44,39 +44,36 @@ def _arrow_to_py_list_scalar(scalar: pa.ListScalar):
 @dataclass
 class ArrowDecoder:
     columns: Dict[str, int]
-    ndarray_columns: Dict[str, int]
 
     @classmethod
     def from_schema(cls, schema):
-        columns = {}
-        ndarray_columns = {}
-
-        for idx, column in enumerate(schema):
-            if column.name.endswith("._np_shape"):
-                ndarray_columns[(column.name.rsplit(".", 1)[0])] = idx
-            else:
-                columns[column.name] = idx
-
-        return cls(columns, ndarray_columns)
+        return cls(
+            [
+                column.name
+                for column in schema
+                if not column.name.endswith("._np_shape")
+            ]
+        )
 
     def decode(self, batch, row_number):
         for row in self.decode_batch(batch.slice(row_number, row_number + 1)):
             return row
 
     def decode_batch(self, batch):
-        rows = zip(*batch)
-
-        for raw_row in rows:
+        for raw_row in batch.to_pandas().to_dict("records"):
             row = {}
-            for column_name, column_idx in self.columns.items():
-                value = _arrow_to_py(raw_row[column_idx])
-
-                shape_idx = self.ndarray_columns.get(column_name)
-
-                if shape_idx is not None:
-                    shape = _arrow_to_py(raw_row[shape_idx])
-                    value = value.reshape(shape)
-
+            for column_name in self.columns:
+                value = raw_row[column_name]
+                shape = raw_row.get(f"{column_name}._np_shape")
+
+                if shape is not None:
+                    value = np.stack(value).reshape(shape)
+                if (
+                    isinstance(value, np.ndarray)
```

Comment:
Couldn't we just check the shape of the array?

Reply:
Hi @jaheba ,

The problem is inline with what we discussed last time. What comes out of the pandas conversion for 2d array is something like

```python
np.array([
    np.array([1,2,3]),
    np.array([4,5,6])
])
```

then if we check the shape, we would only get 1d array. This is why we have 

```python
and len(value.shape) == 1
```

here.

But would be glad to adapt if you have a better idea.

Reply:
To the best of our knowledge this does not work. The problem lies in the pandas conversion of multi-dimensional arrays. One obtains arrays of arrays (instead of multi dim arrays)

Reply:
Oh, I see. Thanks for the clarification.

I think neither arrow nor pandas support arrays with 2 or more dimensions, thus resulting in these nested arrays, with `dtype=object`.

That's why the re-shaping approach makes things so much easier.

Playing around it looks like `np.stack` is indeed able to handle nested arrays just fine, even with dimensions higher than 2.

However, I think we can enable stacking only on arrays which `dtype == object`.